cpp/src/arrow/util/future.h

```diff
@@ -56,6 +47,39 @@ struct is_future<Future<T>> : std::true_type {};
 template <typename Signature>
 using result_of_t = typename std::result_of<Signature>::type;
 
+template <typename Source, typename Dest>
+typename std::enable_if<Source::is_empty>::type Propagate(Source& source, Dest dest) {
```

Comment:
Style nit: mutable references should be avoided
```suggestion
typename std::enable_if<Source::is_empty>::type Propagate(Source* source, Dest dest) {
```

Reply:
Changed to a pointer.